src/main/scala/mesosphere/marathon/core/election/ElectionService.scala

```diff
@@ -91,10 +105,155 @@ trait ElectionCandidate {
   def startLeadership(): Unit
 }
 
-/** Local leadership events. They are not delivered via the event endpoints. */
-sealed trait LocalLeadershipEvent
+class ElectionServiceImpl(
+    eventStream: EventStream,
+    hostPort: String,
+    leaderEventsSource: Source[LeadershipState, Cancellable],
+    crashStrategy: CrashStrategy,
+    electionEC: ExecutionContext
+)(implicit system: ActorSystem) extends ElectionService with StrictLogging {
+
+  import ElectionService._
+  @volatile private[this] var lastState: LeadershipState = LeadershipState.Standby(None)
```

Comment:
What's the advantage of `@volatile` over and atomic reference?

Reply:
AtomicReference would work.

AtomicReference creates one more box, since it is generic.

I generally use AtomicReference when I need to apply atomic operations, such as compare and swap. That isn't needed here. `@volatile` var suffices.

Reply:
`@volatile` is fine 🙂 .